File: src/advent/day3.ts
export function solve(input: Array<string>): [number, number] {
  return [part1(input), part2(input)];
}

export function part1(input: Array<string>): number {
  let total = 0;

  for (const item of input) {
    const [compA, compB] = splitIntoCompartments(item);
    const commonItem = findCommonItem(compA.sort(), compB.sort());

    if (commonItem.length > 0) {
      total += calculatePriority(commonItem);
    }
  }

  return total;
}

export function part2(input: Array<string>): number {
  let total = 0;

  for (let i = 0; i < input.length; i += 3) {
    const group = [
      Array.from(input[i]).sort(),
      Array.from(input[i + 1]).sort(),
      Array.from(input[i + 2]).sort(),
    ];

    const commonItem = findCommonItem(...group);
    if (commonItem.length > 0) {
      total += calculatePriority(commonItem);
    }
  }

  return total;
}

export function splitIntoCompartments(input: string): [string[], string[]] {
  const half = input.length / 2;
  return [Array.from(input.slice(0, half)), Array.from(input.slice(half))];
}

export function findCommonItem(...groups: string[][]): string {
  while (groups.every((g) => g.length > 0)) {
    const firsts = groups.map((g) => g[0]);

    if (firsts.every((f) => f === firsts[0])) {
      return firsts[0];
    }

    const min = Math.min(...firsts.map((f) => f.charCodeAt(0)));
    groups.forEach((g) => {
      if (g[0].charCodeAt(0) === min) {
        g.shift();
      }
    });
  }

  return "";
}

export function calculatePriority(input: string): number {
  const value = input.charCodeAt(0);
  if (value >= 97 && value <= 122) {
    // a-z
    return value - "a".charCodeAt(0) + 1;
  } else {
    // A-Z
    return value - "A".charCodeAt(0) + 27;
  }
}

File: rust/src/day3.rs
use core::cmp::Ordering;

pub fn run() -> (Option<u32>, Option<u32>) {
    let input = include_str!("../data/day3.txt");
    return (part1(input), part2(input));
}

const TINY_A: u32 = 'a' as u32;
const BIG_A: u32 = 'A' as u32;

#[derive(Debug)]
struct Backpack {
    items: Vec<u32>,
    index: usize,
}

fn part1(input: &str) -> Option<u32> {
    let mut total = 0;

    for line in input.split('\n') {
        let mut chars: Vec<u32> = line.chars().map(|c| c as u32).collect();
        let half = chars.len() / 2;
        let (comp_a, comp_b) = chars.split_at_mut(half);

        match find_common_item(&vec![comp_a.to_vec(), comp_b.to_vec()]) {
            Some(value) => total += value,
            None => println!("did not find the common item!"),
        }
    }

    return Some(total);
}

fn part2(input: &str) -> Option<u32> {
    let mut total = 0;

    let all_lines: Vec<&str> = input.split('\n').into_iter().collect();
    let amount = 3;
    for i in (0..all_lines.len()).step_by(amount) {
        let group = all_lines
            .iter()
            .skip(i)
            .take(amount)
            .map(|line| line.chars().map(|c| c as u32).collect())
            .collect();

        match find_common_item(&group) {
            Some(value) => total += value,
            None => println!("did not find the common item!"),
        }
    }

    return Some(total);
}

fn find_common_item(compartments_param: &Vec<Vec<u32>>) -> Option<u32> {
    let mut compartments: Vec<Backpack> = compartments_param
        .iter()
        .map(|c| {
            let mut items = c.to_vec();
            items.sort();
            return Backpack { items, index: 0 };
        })
        .collect();

    while compartments.iter().all(|c| c.index < c.items.len()) {
        let current_items: Vec<u32> = compartments.iter().map(|c| c.items[c.index]).collect();

        let all_items_equal = current_items
            .iter()
            .all(|i| i.cmp(&current_items[0]) == Ordering::Equal);

        if all_items_equal {
            return Some(calculate_priority(current_items[0]));
        }

        let min_value = current_items.iter().min().unwrap();

        for mut compartment in &mut compartments {
            if compartment.items[compartment.index] == *min_value {
                compartment.index += 1;
            }
        }
    }

    return None;
}

fn calculate_priority(value: u32) -> u32 {
    if value >= 97 && value <= 122 {
        return value - TINY_A + 1;
    } else {
        return value - BIG_A + 27;
    }
}

#[cfg(test)]
mod t {
    use super::*;

    const TEST_INPUT: &'static str = "vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw";

    #[test]
    fn test_part1() {
        assert_eq!(157, part1(TEST_INPUT).unwrap());
    }

    #[test]
    fn test_part2() {
        assert_eq!(70, part2(TEST_INPUT).unwrap());
    }

    #[test]
    fn test_find_common_item() {
        let items = vec![vec![97, 101, 113, 100, 98], vec![125, 99, 100, 108, 103]];
        assert_eq!(4, find_common_item(&items).unwrap());
    }

    #[test]
    fn test_not_find_common_item() {
        let items = vec![vec![99], vec![103]];
        assert_eq!(None, find_common_item(&items));
    }
}
